creator-node/src/snapbackSM.js

```diff
@@ -47,63 +58,308 @@ class SnapbackSM {
+    config.set('spID', recoveredSpID)
+  }
+
+  // Retrieve users with this node as primary
+  async getNodePrimaryUsers () {
+    const currentlySelectedDiscProv = this.audiusLibs.discoveryProvider.discoveryProviderEndpoint
+    let requestParams = {
+      method: 'get',
+      baseURL: currentlySelectedDiscProv,
+      url: `users/creator_node`,
+      params: {
+        creator_node_endpoint: this.endpoint
+      }
+    }
+    let resp = await axios(requestParams)
+    this.log(`Discovery provider: ${currentlySelectedDiscProv}`)
+    return resp.data.data
+  }
+
   /*
-    Main state machine processing function
-    TODO: Update job arguments
+    Retrieve the current clock value on this node for the provided user wallet
   */
+  async getUserPrimaryClockValue (wallet) {
+    let walletPublicKey = wallet.toLowerCase()
+    const cnodeUser = await models.CNodeUser.findOne({
+      where: { walletPublicKey }
+    })
+    const clockValue = (cnodeUser) ? cnodeUser.dataValues.clock : -1
+    return clockValue
+  }
+
+  // Enqueue a sync request to a particular secondary
+  async issueSecondarySync (userWallet, secondaryEndpoint, primaryEndpoint) {
+    let syncRequestParameters = {
+      baseURL: secondaryEndpoint,
+      url: '/sync',
+      method: 'post',
+      data: {
+        wallet: [userWallet],
+        creator_node_endpoint: primaryEndpoint,
+        state_machine: true // state machine specific flag
+      }
+    }
+    await this.syncQueue.add({ syncRequestParameters, startTime: Date.now() })
+  }
+
+  // Main state machine processing function
   async processStateMachineOperation (job) {
-    this.log('------------------Process state machine operation------------------')
-    // TODO: Translate working branch replica set processing
-    // First step here is to implement discovery provider query
+    this.log('------------------Process SnapbackSM Operation------------------')
+    if (this.audiusLibs == null) {
+      logger.error(`Invalid libs instance`)
+      return
+    }
+    // Attempt to initialize, if this creator node has not yet been registered on chain
+    // then no operations will be performed
+    if (!this.initialized) {
+      await this.initializeNodeIdentityConfig()
+      return
+    }
+    // 1.) Retrieve base information for state machine operations
+    let spInfo = await this.getSPInfo()
+    if (spInfo.spID === 0) {
+      this.log(`Invalid spID, recovering ${spInfo}`)
+      await this.recoverSpID()
+      return
+    }
+    // Retrieve users list for this node
+    let usersList = await this.getNodePrimaryUsers()
+
+    // Generate list of wallets to query clock number
+    let nodeVectorClockQueryList = {}
+
+    // Issue queries to secondaries for each user
+    await Promise.all(
+      usersList.map(
+        async (user) => {
+          let userWallet = user.wallet
+          let secondary1 = user.secondary1
+          let secondary2 = user.secondary2
```

Comment:
what if `secondary1` or `secondary2` are null?
in that case, looks like it will actually store `nodeVectorClockQueryList[null] = [userWallet]`
i didn't know that but we should add a null check here
guessing you didn't notice since maddog sets both secondaries

Reply:
good catch, all access assumptions should be null-checked within this class